File: frontend/src/features/contract/api/getProjectContract.ts
import { ApiUrl, customFetch, withDelay } from "@/constants/api";

export interface GetProjectContractResponse {
  id: string;
  name: string;
  registrationNumber: string;
  projectStartDate: string;
  projectEndDate: string | null;
}

export const getProjectContract = (projectId: string, contractId: string) =>
  withDelay("fast", async () => {
    try {
      return await customFetch<GetProjectContractResponse>(`${ApiUrl}/projects/${projectId}/contracts/${contractId}`);
    } catch (error) {
      if (error instanceof Error && error.message.startsWith("404")) {
        return null;
      }
      throw error;
    }
  });

File: backend/Timesheets.Api/Features/Projects/Endpoints/GetProjectContract.cs
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.EntityFrameworkCore;
using Timesheets.Api.Data;
using Timesheets.Api.Features.Auth;

namespace Timesheets.Api.Features.Projects.Endpoints;

public sealed class GetProjectContract : IEndpoint
{
    public static void Map(IEndpointRouteBuilder app) =>
        app.MapGet("/{contractEmployeeId}/contracts/{contractId}", Handle)
           .WithSummary("Get Project Contract");

    public sealed record Response(Guid Id, string Name, string RegistrationNumber, DateTime ProjectStartDate, DateTime? ProjectEndDate);

    private static async Task<Results<Ok<Response>, NotFound, ForbidHttpResult>> Handle(Guid contractEmployeeId, Guid contractId, AppDbContext dbContext, ICurrentUser user, CancellationToken cancellationToken)
    {
        if (!user.Satisfies(UserRole.Employee, contractId: contractId))
        {
            return TypedResults.Forbid();
        }

        Response? contract = await dbContext.Contracts
            .AsNoTracking()
            .Where(c => c.ProjectId == contractEmployeeId && c.Id == contractId)
            .Select(c => new Response(
                c.Id,
                c.Name,
                c.RegistrationNumber,
                c.Project.StartDate,
                c.Project.EndDate
            ))
            .FirstOrDefaultAsync(cancellationToken);

        if (contract is null)
        {
            return TypedResults.NotFound();
        }

        return TypedResults.Ok(contract);
    }
}
